File: integrations/mcp_clients/quotabot_mcp_common.py
from __future__ import annotations

import json
import math
import re
from collections.abc import Iterable
from typing import Any
from urllib.parse import urlsplit


_REQUIRED_ROUTING_TOOLS = ("suggest_provider", "suggest_model")
_LOOPBACK_MCP_URL_PATTERN = re.compile(
    r"^https?://(?:localhost|127\.0\.0\.1|\[::1\])"
    r"(?::[0-9]+)?(?:[/?][^\s\\#]*)?$",
    re.IGNORECASE,
)
_LOOPBACK_MCP_HOSTS = {"localhost", "127.0.0.1", "::1"}
_LOOPBACK_MCP_URL_ERROR = (
    "QUOTABOT_MCP_URL must use http or https with the exact loopback host "
    "localhost, 127.0.0.1, or ::1"
)


def require_loopback_mcp_url(value: str) -> str:
    """Validate an MCP HTTP endpoint without resolving alternate host forms."""
    if not isinstance(value, str) or not _LOOPBACK_MCP_URL_PATTERN.fullmatch(value):
        raise ValueError(_LOOPBACK_MCP_URL_ERROR)

    try:
        parsed = urlsplit(value)
        parsed.port
    except ValueError as error:
        raise ValueError(_LOOPBACK_MCP_URL_ERROR) from error

    hostname = parsed.hostname
    if (
        parsed.scheme.lower() not in {"http", "https"}
        or hostname is None
        or hostname.lower() not in _LOOPBACK_MCP_HOSTS
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
    ):
        raise ValueError(_LOOPBACK_MCP_URL_ERROR)
    return value


def structured_content(result: Any) -> dict[str, Any]:
    """Return structured MCP tool content, falling back to JSON text content."""
    direct = getattr(result, "structuredContent", None)
    if isinstance(direct, dict):
        return direct

    snake_case = getattr(result, "structured_content", None)
    if isinstance(snake_case, dict):
        return snake_case

    for item in getattr(result, "content", ()) or ():
        text = getattr(item, "text", None)
        if not isinstance(text, str):
            continue
        try:
            decoded = json.loads(text)
        except json.JSONDecodeError:
            continue
        if isinstance(decoded, dict):
            return decoded

    return {}


def routing_summary(
    suggestion: dict[str, Any],
    model_suggestion: dict[str, Any] | None = None,
) -> dict[str, Any]:
    recommended = _object_value(suggestion.get("recommended"))
    fallback = _object_value(suggestion.get("fallback"))
    model = _object_value((model_suggestion or {}).get("recommended"))

    return {
        "suggest_schema": _string_value(suggestion.get("schema")),
        "recommended_provider": _string_value(recommended.get("provider")),
        "headroom_percent": _number_value(recommended.get("headroom_percent")),
        "using_local_fallback": suggestion.get("using_local_fallback") is True,
        "fallback_provider": _string_value(fallback.get("provider")),
        "model_schema": _string_value((model_suggestion or {}).get("schema")),
        "recommended_model": _string_value(model.get("id")),
        "model_provider": _string_value(model.get("provider")),
    }


def as_pretty_json(value: dict[str, Any]) -> str:
    return json.dumps(value, indent=2, sort_keys=False, allow_nan=False)


def require_routing_tools(tool_names: Iterable[str]) -> None:
    """Raise a clear error when the server lacks a routing tool we call."""
    available = set(tool_names)
    missing = [name for name in _REQUIRED_ROUTING_TOOLS if name not in available]
    if missing:
        raise RuntimeError(f"quotabot MCP tools missing: {', '.join(missing)}")


def _object_value(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _string_value(value: Any) -> str | None:
    return value if isinstance(value, str) and value else None


def _number_value(value: Any) -> int | float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if isinstance(value, float) and not math.isfinite(value):
        return None
    return value

File: integrations/mcp_clients/quotabot_mcp_common.ts
import type { Client } from "@modelcontextprotocol/sdk/client/index.js";

const REQUIRED_ROUTING_TOOLS = ["suggest_provider", "suggest_model"] as const;

export function mcpBearerHeaders() {
  const token = process.env.QUOTABOT_MCP_TOKEN?.trim();
  return token ? { Authorization: `Bearer ${token}` } : undefined;
}

export function structuredContent(result: unknown): Record<string, unknown> {
  const root = objectValue(result);
  if (isRecord(root.structuredContent)) {
    return root.structuredContent;
  }

  const content = root.content;
  if (!Array.isArray(content)) {
    return {};
  }

  for (const item of content) {
    const record = objectValue(item);
    if (typeof record.text !== "string") {
      continue;
    }
    try {
      const decoded = JSON.parse(record.text);
      if (isRecord(decoded)) {
        return decoded;
      }
    } catch {
      continue;
    }
  }

  return {};
}

export function routingSummary(
  suggestion: Record<string, unknown>,
  modelSuggestion: Record<string, unknown> | undefined = undefined,
) {
  const recommended = objectValue(suggestion.recommended);
  const fallback = objectValue(suggestion.fallback);
  const model = objectValue(modelSuggestion?.recommended);

  return {
    suggest_schema: stringValue(suggestion.schema),
    recommended_provider: stringValue(recommended.provider),
    headroom_percent: numberValue(recommended.headroom_percent),
    using_local_fallback: suggestion.using_local_fallback === true,
    fallback_provider: stringValue(fallback.provider),
    model_schema: stringValue(modelSuggestion?.schema),
    recommended_model: stringValue(model.id),
    model_provider: stringValue(model.provider),
  };
}

export async function printRoutingDecision(
  client: Client,
  modelArguments: Record<string, unknown> = {},
): Promise<void> {
  const tools = await client.listTools();
  const names = new Set(tools.tools.map((tool) => tool.name));
  const missing = REQUIRED_ROUTING_TOOLS.filter((name) => !names.has(name));
  if (missing.length > 0) {
    throw new Error(`quotabot MCP tools missing: ${missing.join(", ")}`);
  }

  const suggestion = structuredContent(
    await client.callTool({ name: "suggest_provider", arguments: {} }),
  );
  const model = structuredContent(
    await client.callTool({ name: "suggest_model", arguments: modelArguments }),
  );
  console.log(JSON.stringify(routingSummary(suggestion, model), null, 2));
}

export function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function objectValue(value: unknown): Record<string, unknown> {
  return isRecord(value) ? value : {};
}

function stringValue(value: unknown): string | null {
  return typeof value === "string" && value.length > 0 ? value : null;
}

function numberValue(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}
